src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+            } else if (const IntImm *b = mul->b.as<IntImm>()) {
+                term.coefficient = b->value;
+                term.expr = mul->a;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // Given a summation, extract it into a series of affine terms.
+    std::vector<AffineTerm> extract_summation(const Expr &e) {
+        std::vector<AffineTerm> pending, terms;
+        pending.push_back({e, 1});
+
+        // In case this is not a sum
+        const Add *add = e.as<Add>();
+        const Sub *sub = e.as<Sub>();
+        if (!(add || sub)) {
+            return pending;  // Can't do anything with non-sum.
+        }
+
+        while (!pending.empty()) {
+            AffineTerm next = pending.back();
+            pending.pop_back();
+            const Add *add = next.expr.as<Add>();
+            const Sub *sub = next.expr.as<Sub>();
+            if (add) {
+                pending.push_back({add->a, next.coefficient});
+                pending.push_back({add->b, next.coefficient});
+            } else if (sub) {
+                pending.push_back({sub->a, next.coefficient});
+                pending.push_back({sub->b, -next.coefficient});
+            } else {
+                // Try to recurse, the node might turn into a summation.
+                next.expr = mutate(next.expr);
+
+                if (next.expr.as<Add>() || next.expr.as<Sub>()) {
+                    // After mutation it became an add or sub, throw it back on the pending queue.
+                    pending.push_back(next);
+                } else {
+                    // If this term is a mul by a constant, we can extract out the constant and keep trying.
+                    AffineTerm term;
+                    if (is_mul_by_int_const(next.expr, term)) {
+                        internal_assert(term.expr.defined());
+                        term.coefficient *= next.coefficient;
+                        pending.push_back(term);
+                    } else {
+                        // This is a leaf.
+                        terms.push_back(next);
+                    }
+                }
+            }
+        }
+
+        std::stable_sort(terms.begin(), terms.end(),
+                         [&](const AffineTerm &a, const AffineTerm &b) {
+                             return should_commute(a.expr, b.expr);
+                         });
+
+        return terms;
+    }
+
+    inline std::vector<AffineTerm> simplify_summation(const std::vector<AffineTerm> &terms) {
+        // TODO: for now, only do linear-term simplification.
+        // There are other simplifications that can be done
+        // (i.e. pushing terms into a Min or Max to cancel terms),
+        // but we need a good polynomial-time algorithm for it.
+        return simplify_linear_summation(terms);
+    }
+
+    // Two-finger O(n) algorithm for simplifying sums.
+    std::vector<AffineTerm> simplify_linear_summation(const std::vector<AffineTerm> &terms) {
```

Comment:
I don't understand how this linear-time algorithm works. An explanatory comment would help.

Reply:
added in cf591f763 , let me know if it is satisfactory

Reply:
It's sorted by deep equality comparison? I just see a should_commute. should_commute is only a partial order.

Reply:
No, I think we had a previous discussion about deep_equality being too expensive and should_commute being good enough but we could make it stronger. Should we change that?

Reply:
Let's change to deep equality and see what the impact is.